service/history/ndc/activity_state_replicator.go

```diff
@@ -126,69 +130,109 @@ func (r *ActivityStateReplicatorImpl) SyncActivityState(
 		}
 		return err
 	}
-
-	scheduledEventID := request.GetScheduledEventId()
-	shouldApply, err := r.testVersionHistory(
-		namespaceID,
-		execution.GetWorkflowId(),
-		execution.GetRunId(),
-		scheduledEventID,
+	applied, err := r.syncSingleActivityState(
+		&definition.WorkflowKey{
+			NamespaceID: request.NamespaceId,
+			WorkflowID:  request.WorkflowId,
+			RunID:       request.RunId,
+		},
 		mutableState,
-		request.GetVersionHistory(),
+		&historyservice.ActivitySyncInfo{
+			Version:            request.Version,
+			ScheduledEventId:   request.ScheduledEventId,
+			ScheduledTime:      request.ScheduledTime,
+			StartedEventId:     request.StartedEventId,
+			StartedTime:        request.StartedTime,
+			LastHeartbeatTime:  request.LastHeartbeatTime,
+			Details:            request.Details,
+			Attempt:            request.Attempt,
+			LastFailure:        request.LastFailure,
+			LastWorkerIdentity: request.LastWorkerIdentity,
+			VersionHistory:     request.VersionHistory,
+			BaseExecutionInfo:  request.BaseExecutionInfo,
+		},
 	)
-	if err != nil || !shouldApply {
+	if err != nil {
 		return err
 	}
-
-	activityInfo, ok := mutableState.GetActivityInfo(scheduledEventID)
-	if !ok {
-		// this should not retry, can be caused by out of order delivery
-		// since the activity is already finished
+	if !applied {
 		return nil
 	}
-	if shouldApply := r.testActivity(
-		request.GetVersion(),
-		request.GetAttempt(),
-		timestamp.TimeValue(request.GetLastHeartbeatTime()),
-		activityInfo,
-	); !shouldApply {
-		return nil
+
+	updateMode := persistence.UpdateWorkflowModeUpdateCurrent
+	if state, _ := mutableState.GetWorkflowStateStatus(); state == enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE {
+		updateMode = persistence.UpdateWorkflowModeBypassCurrent
 	}
 
-	// sync activity with empty started ID means activity retry
-	eventTime := timestamp.TimeValue(request.GetScheduledTime())
-	if request.StartedEventId == common.EmptyEventID && request.Attempt > activityInfo.GetAttempt() {
-		mutableState.AddTasks(&tasks.ActivityRetryTimerTask{
-			WorkflowKey: definition.WorkflowKey{
-				NamespaceID: request.GetNamespaceId(),
-				WorkflowID:  request.GetWorkflowId(),
-				RunID:       request.GetRunId(),
-			},
-			VisibilityTimestamp: eventTime,
-			EventID:             request.GetScheduledEventId(),
-			Version:             request.GetVersion(),
-			Attempt:             request.GetAttempt(),
-		})
+	return executionContext.UpdateWorkflowExecutionWithNew(
```

Comment:
not sure if we should check the workflow running state

Reply:
I checked the original implementation, seems we did not check running state. What you mean is that we should make sure workflow is in running state before we UpdateWorkflowExecution?